### PR TITLE
Rewriter Optimization

### PR DESCRIPTION
**Pull request by Kieranoski702**
This PR adds the "clean/dirty" optimisation. The rewriter algorithm and this optimisation are explained in [this issue](https://github.com/conjure-cp/conjure-oxide/issues/149). Clean nodes are skipped as there is no work to be done with them and dirty nodes are rewritten normally. Whenever a node is turned dirty then it's parent will make itself dirty. 

I have written a new test in `rewriter_tests` called `rewrite_solve_xyz_parameterized` which runs the rewriter over increasing "complex" XYZ models. Complex in this instance refers to the number of branches of execution so each version of the test will have an extra XYZ problem in the OR clause. Currently, this test runs from 1 to 4 number of XYZ problems mainly because 5 and above is quite slow - 5 XYZ problems in the OR runs for around 15 seconds for example.  

Additionally, this tests each number of XYZ problems with and without the new optimisation so that we can compare the speed and number of rule application attempts. An info json is output inside the `conjure_oxide/tests` directory for each configuration of the test (so 8 in total for now). 

Here are some graphs that show the performance gains over the unoptimized version and the amount of rule applications saved over the unoptimized version as the number of XYZ problems increases. I have removed the data where there is only one XYZ problem as the results are very very inconsistent probably due to its very low run time. 

performance_gain_1
rule_application_savings_1

It is important to note that the performance increase seems to vary from run to run but is always positive. It seems that on most runs one number of XYZ problems gets quite a large increase in performance but this shifts from run to run - could this be from some sort of caching? Any insight would be appreciated!

I plan to add an option to the tests to increase the number of XYZ problems to a set amount but keep the default at 4 so I can collect some more results without affecting the test runtime. I will also do a few runs and average out the data for some better graphs but I think this shows the general idea :)

### CODE REVIEW

**Comment by ozgurakgun** [comment]
re the graphs: I wouldn't worry about time but instead show the number of rule attempts with and without the optimisation. which is the second plot, right? I am very surprised to see the small difference in savings between 3 and 4, why is that?

instead of percentage, just make a table with 3 columns: number of disjuncts in the or, number of rule attempts without optimisation, number of rule attempts with optimisation? percentage hides some info (like the actual magnitude).

**Comment by Kieranoski702** [comment]
> re the graphs: I wouldn't worry about time but instead show the number of rule attempts with and without the optimisation. which is the second plot, right? I am very surprised to see the small difference in savings between 3 and 4, why is that?
> 
> instead of percentage, just make a table with 3 columns: number of disjuncts in the or, number of rule attempts without optimisation, number of rule attempts with optimisation? percentage hides some info (like the actual magnitude).

I'll take a look at why there is a small difference between 3 and 4 as well as make the table with that direct information instead of the graph for the presentation tomorrow. Thank you!

File: conjure_oxide/tests/rewrite_tests.rs [hunk 1]
```diff
@@ -914,6 +916,154 @@ fn rewrite_solve_xyz() {
     solver.solve(Box::new(|_| true)).unwrap();
 }
 
+#[test]
+fn rewrite_solve_xyz_parameterized() {
+    println!("Rules: {:?}", get_rules());
+
+    let rule_sets = match resolve_rule_sets(SolverFamily::Minion, &vec!["Constant"]) {
```
Comment (ChrisJefferson) [inline conjure_oxide/tests/rewrite_tests.rs:923]: I think line needs a '.to_string()', it broke building from a fresh checkout?

Comment (Kieranoski702) [reply conjure_oxide/tests/rewrite_tests.rs:923]: I am quite surprised that something isn't building on a fresh checkout. I took those lines from the previous rewrite tests which did not seem to break anything and the tests passed all the CI build checks. I have also built this on a lab machine and it seems to build fine. I have only tried manually building on Ubuntu and Fedora though - what OS is this breaking on @ChrisJefferson?

Comment (ozgurakgun) [reply conjure_oxide/tests/rewrite_tests.rs:923]: also fails in CI: https://github.com/conjure-cp/conjure-oxide/actions

Comment (ozgurakgun) [reply conjure_oxide/tests/rewrite_tests.rs:923]: may be better to look at the recent commits actually: https://github.com/conjure-cp/conjure-oxide/commits/main/

Comment (ChrisJefferson) [reply conjure_oxide/tests/rewrite_tests.rs:923]: This is just an occasional issue with github, you can have two PRs (this one, and the simple CLI one) which both pass tests independantly, but don't merge together.

Basically github tests that (a) your branch is mergable, and (b) it passes all tests, but it (annoyingly) doesn't test that your branch will pass all tests once merged! I've always been suprised github doesn't have an option to fix this, but all solutions to this seem to be more work than it occasionally happening...

**Pull request merged**